src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+    size_t actual_index = 0;
+    StateData state(mIndex);
+
+    while (actual_index++ < mTotalCount)
+    {
+        if (CHIP_NO_ERROR != state.Load(mProvider.mStorage))
+        {
+            break;
+        }
+        if (state.fabric_index == mFabric)
+        {
+            count++;
+        }
+        state.index = state.next;
+    }
+    return count;
+}
+
+bool GroupDataProviderImpl::FabricStatesIterator::Next(GroupState & item)
+{
+    StateData state(mIndex);
+
+    while (mCount++ < mTotalCount)
+    {
+        if (CHIP_NO_ERROR != state.Load(mProvider.mStorage))
+        {
+            mCount = mTotalCount;
+            break;
+        }
+        state.index = state.next;
+        if (state.fabric_index == mFabric)
+        {
+            item.fabric_index = state.fabric_index;
+            item.group        = state.group;
+            item.keyset_index = state.keyset_index;
+            mIndex            = state.index;
+            return true;
+        }
+    }
+    return false;
+}
+
+void GroupDataProviderImpl::FabricStatesIterator::Release()
+{
+    mProvider.mFabricStatesIterators.ReleaseObject(this);
+}
+
+//
+// Key Sets
+//
+
+CHIP_ERROR GroupDataProviderImpl::SetKeyset(chip::FabricIndex fabric_index, uint16_t target_id, const Keyset & in_keyset)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+
+    FabricData fabric(fabric_index);
+    if (CHIP_NO_ERROR != fabric.Load(mStorage))
+    {
+        // First in_keyset
+        fabric.keyset_count = 1;
+        fabric.first_keyset = target_id;
+        KeysetData keyset(fabric_index, target_id, in_keyset.policy, in_keyset.num_keys_used);
+        memcpy(keyset.epoch_keys, in_keyset.epoch_keys, sizeof(in_keyset.epoch_keys));
+        ReturnLogErrorOnFailure(keyset.Save(mStorage));
+        return fabric.Save(mStorage);
+    }
+
+    // Pseudo-index > 0
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    size_t keyset_count = 0;
+    bool found          = false;
+
+    // Seach for the target in_keyset id
+    while (keyset_count < fabric.keyset_count)
+    {
+        if (CHIP_NO_ERROR != keyset.Load(mStorage))
+        {
+            break;
+        }
+        if (keyset.keyset_id == target_id)
+        {
+            // Target id found
+            found = true;
+            break;
+        }
+        keyset.keyset_id = keyset.next;
+        keyset_count++;
+    }
+
+    keyset.keyset_id     = target_id;
+    keyset.policy        = in_keyset.policy;
+    keyset.num_keys_used = in_keyset.num_keys_used;
+    memcpy(keyset.epoch_keys, in_keyset.epoch_keys, sizeof(in_keyset.epoch_keys));
+
+    if (found)
+    {
+        // Update existing in_keyset info, keep next
+        return keyset.Save(mStorage);
+    }
+
+    // New in_keyset, insert first
+    keyset.next = fabric.first_keyset;
+    ReturnErrorOnFailure(keyset.Save(mStorage));
+    // Update fabric info
+    fabric.keyset_count++;
+    fabric.first_keyset = target_id;
```

Comment:
Why are we changing first_keyset here?

Reply:
Because new entries are inserted first. I used to insert the new entries last, but as @bzbarsky-apple pointed-out, since this is a linked-list, inserting first is easier and faster.